tracer/src/Datadog.Trace/Debugger/Upload/SymbolUploadApi.cs

```diff
@@ -0,0 +1,115 @@
+        private const int StartingSleepDuration = 3;
+
+        private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor<SymbolUploadApi>();
+
+        private readonly IApiRequestFactory _apiRequestFactory;
+        private readonly ArraySegment<byte> _eventMetadata;
+
+        private SymbolUploadApi(
+            IApiRequestFactory apiRequestFactory,
+            IDiscoveryService discoveryService,
+            IGitMetadataTagsProvider gitMetadataTagsProvider,
+            ArraySegment<byte> eventMetadata)
+            : base(apiRequestFactory, gitMetadataTagsProvider)
+        {
+            _apiRequestFactory = apiRequestFactory;
+            _eventMetadata = eventMetadata;
+
+            discoveryService.SubscribeToChanges(c => Endpoint = c.SymbolDbEndpoint);
+        }
+
+        public static IBatchUploadApi Create(
+            IApiRequestFactory apiRequestFactory,
+            IDiscoveryService discoveryService,
+            IGitMetadataTagsProvider gitMetadataTagsProvider,
+            string serviceName)
+        {
+            ArraySegment<byte> GetEventMetadataAsArraySegment()
+            {
+                var sb = new StringBuilder();
+                sb.Append(@"{");
+                sb.Append(@"""ddsource"": ""dd_debugger"",");
+                sb.Append(@$"""service"": ""{serviceName}"",");
+                sb.Append(@$"""runtimeId"": ""{Tracer.RuntimeId}""");
+                sb.Append(@"}");
+                var eventMetadata = sb.ToString();
+
+                var count = Encoding.UTF8.GetByteCount(eventMetadata);
+                var eventAsBytes = new byte[count];
+                Encoding.UTF8.GetBytes(eventMetadata, 0, eventMetadata.Length, eventAsBytes, 0);
+                return new ArraySegment<byte>(eventAsBytes);
+            }
+
+            var eventMetadata = GetEventMetadataAsArraySegment();
+            return new SymbolUploadApi(apiRequestFactory, discoveryService, gitMetadataTagsProvider, eventMetadata);
+        }
+
+        public override async Task<bool> SendBatchAsync(ArraySegment<byte> symbols)
+        {
+            var uri = BuildUri();
+            if (string.IsNullOrEmpty(uri))
+            {
+                Log.Warning("Symbol database endpoint is not defined");
+                return false;
+            }
+
+            var request = _apiRequestFactory.Create(new Uri(uri));
+            var multipartRequest = (IMultipartApiRequest)request;
+
+            var retries = 0;
+            var sleepDuration = StartingSleepDuration;
+
+            while (retries < MaxRetries)
+            {
+                using var response =
+                    await multipartRequest
+                         .PostAsync(new MultipartFormItem[]
+                          {
+                              new("file", MimeTypes.Json, "file.json", symbols),
+                              new("event", MimeTypes.Json, "event.json", _eventMetadata)
+                          })
```

Comment:
Do you want to keep recreating this array in the while loop? Can we move it outside?

Reply:
No I don't 😅 
Fixed here 98463d56df340bd1a02fa17c588e50df644f7d99